### ISSUE TITLE
Crash in decoded(scale:)

### ISSUE THREAD

**Issue by Standret** [onevcat/Kingfisher#1381]
### Issue Description

I've encountered this problem when the app was running in the background, and this issue was reported from users in crashlytics.

#### Reproduce

```
Kingfisher                     0x101de3760 specialized Kingfisher<A>.decoded(scale:) + 751 (Image.swift:751)
Kingfisher                     0x101de81e8 closure #1 in closure #3 in ImageCache.retrieveImage(forKey:options:completionHandler:) + 724 (Image.swift:724)
Kingfisher                     0x101de56a4 thunk for @escaping @callee_guaranteed () -> () + 4392113828 (<compiler-generated>:4392113828)
```

This exception occurs inside `processImage (604) ImageDownloader.swift` and crashed in Kingfisher<A>.decoded(scale:) + 751 with attempt unwrap nil value (makeImage from CGContext returns nil)
Unfortunately, I can't provide you with the URL. because it stores on the local device.

**Comment by onevcat** [comment]
:( May I know the version of Kingfisher you are using? It seems that your log is not symbolized correctly. Or it is a very old version of Kingfisher, because the symbols are not existing in the current version...

**Comment by Standret** [comment]
@onevcat I use 4.10.1

**Comment by onevcat** [comment]
Maybe it is worth to try upgrade to Kingfisher 5.x, it brings a defensive memory cache usage may help your case. Or you can try to limit the max memory usage like this: https://github.com/onevcat/Kingfisher/wiki/Cheat-Sheet#set-limit-for-cache

**Comment by Standret** [comment]
Thanks for the advice, I'll try to migrate the project to new version hope it solves the problem.

I suppose currently we can close this issue if I discover a similar one with the new version I'll inform you here.

**Comment by SwifterCat** [comment]
Can the new version of the SDK solve this issue? I also encountered the same problem as you.

**Comment by wow-such-amazing** [comment]
Hey :) 

I think I have the same crash occurring and the version I use is `6.2.1`. I will try to create an example project to replicate this issue. In the meantime I attach a screenshot. Maybe it can help to identify the problem. I added a breakpoint in `malloc_error_break` to execute some actions in the console. They are also on the screenshot, though useless I think 😄

Also, I use AnimatedImageView and I load several quite heavy GIFs. 

![Screenshot 2021-04-13 at 18 41 28](https://user-images.githubusercontent.com/7022709/114589340-de5bac80-9c87-11eb-8f2d-6441d19e1b83.png)

**Comment by wow-such-amazing** [comment]
@onevcat I was able to reproduce an issue with this simple setup. Unfortunately I don't think I can use my gif url, cause it's related to the client. And I don't know how to google a large gif with many frames. For info the gif I was using is 25 mb large and has 350 frames. I will try to find some public gif that could be used to reproduce this bug. 

```
import UIKit
import Kingfisher

class ViewController: UIViewController {
    let stackView = UIStackView()
    let imageView_1 = AnimatedImageView()
    let imageView_2 = AnimatedImageView()
    let imageView_3 = AnimatedImageView()
    let imageView_4 = AnimatedImageView()

    override func viewDidLoad() {
        super.viewDidLoad()

        stackView.translatesAutoresizingMaskIntoConstraints = false
        
        view.addSubview(stackView)
        
        NSLayoutConstraint.activate([
            stackView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
            stackView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            stackView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
            stackView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor),
        ])
        
        stackView.axis = .vertical
        stackView.distribution = .fillEqually
        
        stackView.addArrangedSubview(imageView_1)
        stackView.addArrangedSubview(imageView_2)
        stackView.addArrangedSubview(imageView_3)
        stackView.addArrangedSubview(imageView_4)
        
        imageView_1.contentMode = .scaleAspectFit
        imageView_2.contentMode = .scaleAspectFit
        imageView_3.contentMode = .scaleAspectFit
        imageView_4.contentMode = .scaleAspectFit

        let url = URL(string: "<large_gif_url>")
        
        imageView_1.kf.setImage(with: url)
        imageView_2.kf.setImage(with: url)
        imageView_3.kf.setImage(with: url)
        imageView_4.kf.setImage(with: url)
    }
}
```

**Comment by wow-such-amazing** [comment]
I finally found a way to have it reproduced! I recorded my screen and turned this into GIF :D 

GIF itself is only 1.7 mb, but has 619 frames. So here is a project to reproduce. It doesn't crash for me in the simulator, but it crashes when I launch the project on device. In my case I use iPhone X for testing. Also it doesn't happen instantly, I need to wait a bit to have it crashed.

[TheCrash.zip](https://github.com/onevcat/Kingfisher/files/6309423/TheCrash.zip)

**Comment by wow-such-amazing** [comment]
I did a small change which helped to avoid crash and also it improved the CPU usage. This change is just a proof of concept. Though I'm not sure if that's the right direction or how it could be used. Hopefully, I will investigate it more. 

So the change I did was that `UIGraphicsBeginImageContextWithOptions` is now called only once. And I also commented the `UIGraphicsEndImageContext`. Looks like constant Begin/End Context increases CPU usage. And also it looks like at some point these calls lead to some memory corruption. I guess due to the frequency of these calls they become non-synchronised at some point 🤔 

So maybe we can call `UIGraphicsBeginImageContextWithOptions` only when we start GIF in AnimatedImageView and we can end context when we stop GIF in AnimatedImageView 🤔
